iwyu.cc

```diff
@@ -1375,62 +1375,71 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   SourceLocation GetCanonicalUseLocation(SourceLocation use_loc,
                                          const NamedDecl* decl) {
     // If we're not in a macro, just echo the use location.
-    if (!IsInMacro(use_loc))
+    if (!use_loc.isMacroID())
       return use_loc;
 
-    // If the macro definition file contains a forward-declaration for decl,
-    // we treat that as a use-attribution hint. See below for details.
-    const FileEntry* macro_def_file = GetFileEntry(GetSpellingLoc(use_loc));
+    clang::SourceManager* sm = GlobalSourceManager();
+
+    VERRS(5) << "Trying to determine use location for '"
+             << PrintableDecl(decl) << "'\n";
+
+    // Spin over the expansions from the inside out, looking for the first
+    // expansion location (i.e. not concatenated or part of a macro argument.)
+    // Inspired by clang's DiagnosticRenderer::emitMacroExpansions.
+    SourceLocation loc = use_loc;
+    while (loc.isMacroID() && sm->isMacroArgExpansion(loc) &&
```

Comment:
Looking at this now, this must be wrong -- should be `sm->isMacroArgExpansion(loc) || IsInScratchSpace(loc)`.

This passes all the tests, hopefully the fixed-up version will, too.